dp_consistency/README.md

```diff
@@ -1,54 +1,112 @@
-   - **Zero-Variance Constraints:** Measurements with zero variance are treated as exact values, and equality constraints are added accordingly.
+### Cover Relationships ###
 
-3. **Quadratic Programming:**
-   - The objective function is the variance-adjusted L2 norm, which acts as a maximum likelihood estimate under the assumption of Gaussian noise.
-   - The `qpsolvers` library is used to solve the QP problem, finding the optimal solution that minimizes inconsistencies while respecting the constraints.
+For each metric and period, it identifies the set representing the union of all entities/dimensions (EDPs) and the
+individual
+_sets for each EDP and add a "cover" relationship to the spec, indicating that the union set's measurement should be_
+greater than or equal to the sum of the measurements of its constituent EDP sets.
 
-## Usage Example
+We do the same for the unions computed for unique reach purposes. Also including the relationship that the
+of all-but-edp-x set and the set x are a cover of the union of all EDPs.
 
-```python
-from collections import defaultdict
-# ... (Your classes: Measurement, SetMeasurementsSpec)
-from qpsolvers import solve_qp
-import numpy as np
+### Subset Relationships ###
 
+For each metric, period, and EDP, it establishes a subset relationship between the set representing that specific EDP
+and the union set for that metric and period, for all unions that include that EDP for tha metric and period.
 
-# ... (Load or create your data)
+If there's a predefined hierarchy among metrics (where some metrics are inherently greater than or equal to others), it
+enforces this relationship for each period and EDP, including the measured unions.
 
-# Create a SetMeasurementsSpec and populate it
-spec = SetMeasurementsSpec()
-spec.add_subset_relation(1, 2)  # Set 2 is a subset of set 1
-spec.add_cover(3, [4, 5])       # Sets 4 and 5 form a cover of set 3
-# ... (Add measurements for each set)
+It assumes measurements in one period are less than or equal to those in the subsequent period. It adds
+subset relationships to reflect this for each metric and EDP, including the union set.
+
+## Implementation
+
+The methodology is implemented using Python and the `qpsolvers` library to solve the QP problem.
+The `SetMeasurementsSpec` class is used to efficiently store and manage the relationships between sets and their
+measurements, facilitating the automatic generation of constraints for the optimization.
+
+## Experimental Results
+
+Simulation results demonstrate a reduction in variance across all variables in the report, indicating improved
+consistency after applying this method. More details will be added soon.
+
+## Key Advantages
+
+* **Statistically Sound:** Based on MLE, providing a principled approach to handling noisy data.
+* **Flexible:** Can accommodate various types of constraints and relationships between sets.
+* **Efficient:** Leverages state of the art QP solvers for fast and accurate solutions.
```

Comment:
Do we have an idea of scalability as the number of constraints increases?

Reply:
qpsolvers is an interface over many QP implementations: they have benchmarks here for different algorithms and implementations. 

https://github.com/qpsolvers/qpbenchmark

currently we are using claralabel which uses interior point which I think is it takes roughly ~O(n^3) in number of variables and constants. I think the # of constants can be affected by the number of constraints, but I have to dig deeper.

I've been doing some benchmarks and seems to behave well for 4 EDPs, more benchmarking results coming soon.

Reply:
Can you add a summary of the results in a follow-up PR when you have them?